pkg/logs/internal/util/container_or_k8s.go

```diff
@@ -0,0 +1,48 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package util
+
+import "github.com/DataDog/datadog-agent/pkg/config"
+
+// LogWhat is the answer to ContainersOrPods
+type LogWhat int
+
+const (
+	// LogContainers means that the logs-agent should log containers, not pods.
+	LogContainers LogWhat = iota
+
+	// LogPods means that the logs-agent should log pods, not containers.
+	LogPods
+
+	// LogNothing means neither containers nor pods are supported.
+	LogNothing
+)
+
+// ContainersOrPods determines how the logs-agent should handle
+// containers: either monitoring individual containers, or monitoring pods and
+// logging the containers within them.
+func ContainersOrPods() LogWhat {
+	c := config.IsFeaturePresent(config.Docker) ||
+		config.IsFeaturePresent(config.Containerd) ||
+		config.IsFeaturePresent(config.Cri) ||
+		config.IsFeaturePresent(config.Podman)
+	k := config.IsFeaturePresent(config.Kubernetes)
+
+	switch {
+	case c && !k:
+		return LogContainers
+	case k && !c:
+		return LogPods
+	case k && c:
+		// prefer kubernetes if k8s_container_use_file is set
+		if config.Datadog.GetBool("logs_config.k8s_container_use_file") {
```

Comment:
On @olivielpeau 's comment. I think the change in behavior here is that previously we relied on both the feature being enabled AND the util (docker or K8s) being available where the util would retry. 

IIRC the core case https://github.com/DataDog/datadog-agent/pull/8148 was trying to solve was that both docker an K8s were available as features. But the agent would only (forever) retry docker if the kublet failed to connect once. 

With this PR - I think the agent will make the correct decision based on the config flags and feature availability, but will not dynamically failover if one of the utils is unavailable. I think the bigger question is - what scenarios rely specifically on the failover behavior? I think in a perfect world an agent configuration should decide what features it expects to use explicitly. But I think we should avoid the possibility of breaking existing agents.